### ISSUE TITLE
Is the test phase allowed for gradient computing and backpropagation?

### ISSUE THREAD

**Issue by ginobilinie** [BVLC/caffe#3752]
Hi,

I'm a little confused about test phase. In some prototxts, test phase use softmax_loss or euclideanloss, as we know, softmax_loss and euclideanloss will involve gradient computing and backward propagation. In this way, the test examples involves optimizing the network. It seems wrong. Is the test phase not allowed to use softmax_loss (it should only use softmax and accuracy which not involves graident backpropagtion)?

Thanks.

**Comment by seanbell** [comment]
Sure, you can compute any loss function (including softmax etc) on test examples and that's a perfectly valid thing to do.  When you do that, caffe will compute the loss value (forwards pass) but not the gradient (backwards pass).  Whether or not you do backpropagation and the type of loss function are totally separate concepts.  In the future, please ask questions like this on the mailing list.

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by ginobilinie** [comment]
@seanbell 

Thank you.
I'll ask modeling/usage questions on user-list in the future.
